Adds user-selectable hash merging support in vars

leucos opened the pull request:
Since Michael is open to a patch that doesn't change Ansible current behaviour, here is my attempt.

Hash variables are currently overriden if they are redefined. This
doesn't let the user refine hash entries or overriding selected keys,
which can, for some, be a desirable feature.
This patch let the user force hash merging by setting the
hash_behaviour value to "merge" (without the quotes) in ansible.cfg

However, by default, ansible behaves like it always did and if any value
besides "merge" is used ("replace" is suggested in the example ansible.cfg
file), it will also behave as always.

As always, triple check my python...

Thanks for your time.

leucos committed:
Adds user-selectable hash merging support in vars

Hash variables are currently overriden if they are redefined. This
doesn't let the user refine hash entries or overriding selected keys,
which can, for some, be a desirable feature.
This patch let the user force hash merging by setting the
hash_behaviour value to "merge" (without the quotes) in ansible.cfg

However, by default, ansible behaves like it always did and if any value
besides "merge" is used ("replace" is suggested in the example ansible.cfg
file), it will also behave as always.

dhozac commented:
Why deepcopy? Vars are essentially immutable once loaded, there should not be any modifications. Have you seen anything to the contrary?

leucos commented:
Indeed, this version works too : 

```
def merge_hash(a, b):
      ''' merges hash b into a
      this means that if b has key k, the resulting has will have a key k
      which value comes from b
      said differently, all key/value combination from b will override a's '''

      # and iterate over b keys
      for k, v in b.iteritems():
          if k in a and isinstance(a[k], dict):
              # if this key is a hash and exists in a
              # we recursively call ourselves with·
              # the key value of b
              a[k] = merge_hash(a[k], v)
          else:       
              # k is not in a, no need to merge b, we just copy
              # or k is not a dictionnary, no need to merge b either, we just copy it
              a[k] = v                                                                                                                        
      # finally, return the resulting hash when we're done iterating keys
      return a        
```

Should I push this in the PR ?

dhozac commented:
I would prefer that.

leucos committed:
Fixes hash merging

No need for deep copy (vars are essentially immutable once loaded, there
should not be any modifications)

mpdehaan commented:
Yes, please.   I would _think_ the memory implications of the other may be
bad for us when targetting a large number of hosts.

leucos commented:
Yes you're right. This has been fixed in 8eb7d74

dhozac commented:
This works for me. Merged, thanks!

leucos commented:
Great ! Thanks a lot guys.
